### PR TITLE
A0-1921: Add merge relation

### PR DESCRIPTION
**Pull request by maciejzelaszczyk**
# Description

Introduces the _merge_ relation for the pooling of two notes.

## Type of change
- New feature (non-breaking change which adds functionality)

# Checklist:
- I have added tests
- I have created new documentation

### CODE REVIEW

File: relations/src/shielder/merge.rs [hunk 1]
```diff
@@ -0,0 +1,319 @@
+use liminal_ark_relation_macro::snark_relation;
+
+/// It expresses the facts that:
+///  - `first_old_note` is a prefix of the result of hashing together the `token_id`,
```
Comment (pmikolajczyk41) [inline relations/src/shielder/merge.rs:4]: "prefix of ..." is no more - this is just a hash

File: relations/src/shielder/merge.rs [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        let merkle_root = FpVar::new_input(ns!(cs, "merkle root"), || self.merkle_root())?;
+        let first_path_shape = PathShapeVar::new_witness(ns!(cs, "first path shape"), || {
+            Ok((*self.max_path_len(), self.first_leaf_index().cloned()))
+        })?;
+
+        check_merkle_proof(
+            merkle_root.clone(),
+            first_path_shape,
+            first_old_note,
+            self.first_merkle_path().cloned().unwrap_or_default(),
```
Comment (pmikolajczyk41) [inline relations/src/shielder/merge.rs:168]: maybe let's use `unwrap()` - the path should never be empty, and if it is, we should panic

File: relations/src/shielder/merge.rs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+            second_leaf_index,
+            first_old_note,
+            second_old_note,
+            first_old_token_amount,
+            second_old_token_amount,
+            new_token_amount,
+        )
+    }
+
+    #[test]
```
Comment (pmikolajczyk41) [inline relations/src/shielder/merge.rs:287]: we don't have negative testcases - maybe we should add some of them to ensure that every of the 4 checks is actually happening

**Review by pmikolajczyk41 — APPROVED** [review]
nice!

File: relations/src/shielder/merge.rs (hunk 1 above)
Comment (maciejzelaszczyk) [reply relations/src/shielder/merge.rs]: OK, good catch, changed.

**Review by mike1729 — APPROVED** [review]

File: relations/src/shielder/merge.rs (hunk 2 above)
Comment (maciejzelaszczyk) [reply relations/src/shielder/merge.rs:168]: This actually not true in our tests so far ;) Key generation uses an empty circuit for which the constraints are generated here https://github.com/Cardinal-Cryptography/aleph-node/blob/0af5afb4356829d56425e2dac16c7a7253141762/relations/src/shielder/deposit_and_merge.rs#L221. This is the likely reason why @mike1729 went for `unwrap_or_default`. The way I see it, we can leave the default version and add additional checks in the tests. I am open to other suggestions, though.

**Review by pmikolajczyk41 — APPROVED** [review]
good to go, thanks!

**Pull request merged**